<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Run the MEGAHIT assembler &mdash; 2017-dibsi-metagenomics 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinxcontrib-images/LightBox2/lightbox2/css/lightbox.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2/js/lightbox.min.js"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2-customize/jquery-noconflict.js"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="2017-dibsi-metagenomics 1.0 documentation" href="toc.html" />
    <link rel="next" title="Annotation with Prokka" href="prokka_tutorial.html" />
    <link rel="prev" title="Short read quality and trimming" href="quality.html" />


<style type="text/css">
<!-- github-based editing disabled; set github_base_account
     and github_project -->
#editor-trap.toggled {
  display: none;
}
</style>



  </head>
  <body role="document">

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="toc.html">
          2017-dibsi-metagenomics</a>
        <span class="navbar-text navbar-version pull-left"><b>1.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="toc.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">2017 DIBSI Metagenomics Workshop at UC Davis</a></li>
<li class="toctree-l1"><a class="reference internal" href="welcome.html">Welcome!</a></li>
<li class="toctree-l1"><a class="reference internal" href="jetstream/boot.html">Booting a Jetstream Computer Instance for your use!</a></li>
<li class="toctree-l1"><a class="reference internal" href="command-line.html">Intro to Shell Lesson</a></li>
<li class="toctree-l1"><a class="reference internal" href="command-line.html#the-shell">The Shell</a></li>
<li class="toctree-l1"><a class="reference internal" href="command-line.html#for-future-reference">For Future Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="command-line.html#finding-files">Finding files</a></li>
<li class="toctree-l1"><a class="reference internal" href="running-command-line-blast.html">Running command-line BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality.html">Short read quality and trimming</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Run the MEGAHIT assembler</a></li>
<li class="toctree-l1"><a class="reference internal" href="prokka_tutorial.html">Annotation with Prokka</a></li>
<li class="toctree-l1"><a class="reference internal" href="sourmash.html">A sourmash tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="sourmash.html#k-mers-k-mer-specificity-and-comparing-samples-with-k-mer-jaccard-distance">K-mers, k-mer specificity, and comparing samples with k-mer Jaccard distance.</a></li>
<li class="toctree-l1"><a class="reference internal" href="binning.html">Binning a Metagenomic Assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="salmon_tutorial.html">Gene Abundance Estimation with Salmon</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="slice.html">Slicing and dicing with k-mers</a></li>
<li class="toctree-l1"><a class="reference internal" href="anvio.html">Using Anvi&#8217;o to Knit Everything Together</a></li>
<li class="toctree-l1"><a class="reference internal" href="circos_tutorial.html">Using and Installing Circos</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Workflow and repeatability discussion</a></li>
<li class="toctree-l1"><a class="reference internal" href="DATA.html">Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="INSTALL.html">Building this site on your own</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html">2017 / April / Metagenomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="day2-install.html">Day 2 - installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="kmer_trimming.html">K-mer Spectral Error Trimming</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Run the MEGAHIT assembler</a><ul>
<li><a class="reference internal" href="#while-the-assembly-runs">While the assembly runs...</a></li>
<li><a class="reference internal" href="#after-the-assembly-is-finished">After the assembly is finished</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="quality.html" title="Previous Chapter: Short read quality and trimming"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; Short read qu...</span>
    </a>
  </li>
  <li>
    <a href="prokka_tutorial.html" title="Next Chapter: Annotation with Prokka"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">Annotation wi... &raquo;</span>
    </a>
  </li>
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/assemble.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">

    <div class="col-md-12 content">
      
  
  <div class="section" id="run-the-megahit-assembler">
<h1>Run the MEGAHIT assembler<a class="headerlink" href="#run-the-megahit-assembler" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://github.com/voutcn/megahit">MEGAHIT</a> is a very fast, quite
good assembler designed for metagenomes.</p>
<p>First, install it:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>cd ~/
git clone https://github.com/voutcn/megahit.git
cd megahit
make
</pre></div>
</div>
<p>Now, download some data:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>mkdir ~/data
cd ~/data
curl -O https://s3-us-west-1.amazonaws.com/dib-training.ucdavis.edu/metagenomics-scripps-2016-10-12/SRR1976948.abundtrim.subset.pe.fq.gz
curl -O https://s3-us-west-1.amazonaws.com/dib-training.ucdavis.edu/metagenomics-scripps-2016-10-12/SRR1977249.abundtrim.subset.pe.fq.gz
</pre></div>
</div>
<p>These are data that have been run through k-mer abundance trimming
(see <a class="reference internal" href="kmer_trimming.html"><em>K-mer Spectral Error Trimming</em></a>) and subsampled so that we can run an assembly
in a fairly short time period :).</p>
<hr class="docutils" />
<p>Now, finally, run the assembler!</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>mkdir ~/assembly
cd ~/assembly
ln -fs ../data/*.subset.pe.fq.gz .

~/megahit/megahit --12 SRR1976948.abundtrim.subset.pe.fq.gz,SRR1977249.abundtrim.subset.pe.fq.gz \
    -o combined
</pre></div>
</div>
<p>This will take about 15 minutes; at the end you should see output like
this:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="o">...</span> <span class="mi">7713</span> <span class="n">contigs</span><span class="p">,</span> <span class="n">total</span> <span class="mi">13168567</span> <span class="n">bp</span><span class="p">,</span> <span class="nb">min</span> <span class="mi">200</span> <span class="n">bp</span><span class="p">,</span> <span class="nb">max</span> <span class="mi">54372</span> <span class="n">bp</span><span class="p">,</span> <span class="n">avg</span> <span class="mi">1707</span> <span class="n">bp</span><span class="p">,</span> <span class="n">N50</span> <span class="mi">4305</span> <span class="n">bp</span>
<span class="o">...</span> <span class="n">ALL</span> <span class="n">DONE</span><span class="o">.</span> <span class="n">Time</span> <span class="n">elapsed</span><span class="p">:</span> <span class="mf">899.612093</span> <span class="n">seconds</span>
</pre></div>
</div>
<p>The output assembly will be in <code class="docutils literal"><span class="pre">combined/final.contigs.fa</span></code>.</p>
<div class="section" id="while-the-assembly-runs">
<h2>While the assembly runs...<a class="headerlink" href="#while-the-assembly-runs" title="Permalink to this headline">¶</a></h2>
<p>Discuss CAMI paper.</p>
<p>What does, and doesn&#8217;t, assemble?</p>
<p>How good is assembly anyway?</p>
<p>Discussion:</p>
<p>Why would we assemble, vs looking at raw reads?  What are the
advantages and disadvantages?</p>
<p>What are the technology tradeoffs between Illumina HiSeq, Illumina
MiSeq, and PacBio? (Also see <a class="reference external" href="http://ivory.idyll.org/blog/2015-sharon-paper.html">this paper</a>.)</p>
<p>What kind of experimental design considerations should you have if you
plan to assemble?</p>
<p>Some figures: the first two come from work by Dr. Sherine Awad on
analyzing the data from Shakya et al (2014).  The third comes from
an analysis of read search vs contig search of a protein database.</p>
<a class=""
               data-lightbox="group-0116310c-168d-49d8-bc31-a924f6393aa0"
               href="_images/assembler-runtimes.png"
               title=""
               data-title=""
               ><img src="_images/assembler-runtimes.png"
                     class=""
                     width="20%"
                     height="auto"
                     alt=""/>
                </a><a class=""
               data-lightbox="group-6103ac03-d858-4d1a-9d05-acac900d9dc0"
               href="_images/assembler-mapping.png"
               title=""
               data-title=""
               ><img src="_images/assembler-mapping.png"
                     class=""
                     width="20%"
                     height="auto"
                     alt=""/>
                </a><a class=""
               data-lightbox="group-1e40d28a-ae3e-4807-b880-45d6ec71493d"
               href="_images/read-vs-contig-alignment.png"
               title=""
               data-title=""
               ><img src="_images/read-vs-contig-alignment.png"
                     class=""
                     width="20%"
                     height="auto"
                     alt=""/>
                </a></div>
<div class="section" id="after-the-assembly-is-finished">
<h2>After the assembly is finished<a class="headerlink" href="#after-the-assembly-is-finished" title="Permalink to this headline">¶</a></h2>
<p>Let&#8217;s first take a look at the assembly:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>less combined/final.contigs.fa
</pre></div>
</div>
<p>Now we can run a few stats on our assembly. To do this we will use <a class="reference external" href="http://quast.sourceforge.net/quast">QUAST</a></p>
<div class="highlight-python"><div class="highlight"><pre><span></span>cd ~/
git clone https://github.com/ablab/quast.git -b release_4.5
export PYTHONPATH=$(pwd)/quast/libs/
</pre></div>
</div>
<p>Now, run QUAST on the assembly:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>cd ~/assembly
~/quast/quast.py combined/final.contigs.fa -o combined-report
cat combined-report/report.txt
</pre></div>
</div>
<p>What does this say about our assembly? What do the stats <em>not</em> tell us?</p>
<p>At this point we can do a bunch of things:</p>
<ul class="simple">
<li>annotate the assembly (<a class="reference internal" href="prokka_tutorial.html"><em>Annotation with Prokka</em></a>);</li>
<li>evaluate the assembly&#8217;s inclusion of k-mers and reads;</li>
<li>set up a BLAST database so that we can search it for genes of interest;</li>
<li>quantify the abundance of the contigs or genes in the assembly, using the original read data set (<a class="reference internal" href="salmon_tutorial.html"><em>Gene Abundance Estimation with Salmon</em></a>);</li>
<li>bin the contigs in the assembly into species bins;</li>
</ul>
<hr class="docutils" />
<p>Next: <a class="reference internal" href="sourmash.html"><em>A sourmash tutorial</em></a></p>
</div>
</div>



<hr>

<font size="-1"><b>LICENSE:</b>
This documentation and all textual/graphic site content is released
under <a href='http://creativecommons.org/publicdomain/zero/1.0/'>
Creative Commons - 0
(CC0)</a> -- <a href='https://github.com/ngs-docs/2017-ucsc-metagenomics'>fork @
github</a>. </font>

<hr>


<!-- disqus commenting disabled; set disqus_shortname -->

    </div>
    

  </div>
</div>

<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2017, Lab for Data Intensive Biology.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.<br/>
    </p>
  </div>
</footer>




<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', '', 'auto');
  ga('send', 'pageview');

</script>


  </body>
</html>