<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>A sourmash tutorial &mdash; 2017-dibsi-metagenomics 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinxcontrib-images/LightBox2/lightbox2/css/lightbox.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2/js/lightbox.min.js"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2-customize/jquery-noconflict.js"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="2017-dibsi-metagenomics 1.0 documentation" href="toc.html" />
    <link rel="next" title="Binning a Metagenomic Assembly" href="binning.html" />
    <link rel="prev" title="Annotation with Prokka" href="prokka_tutorial.html" />


<style type="text/css">
<!-- github-based editing disabled; set github_base_account
     and github_project -->
#editor-trap.toggled {
  display: none;
}
</style>



  </head>
  <body role="document">

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="toc.html">
          2017-dibsi-metagenomics</a>
        <span class="navbar-text navbar-version pull-left"><b>1.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="toc.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">2017 DIBSI Metagenomics Workshop at UC Davis</a></li>
<li class="toctree-l1"><a class="reference internal" href="welcome.html">Welcome!</a></li>
<li class="toctree-l1"><a class="reference internal" href="jetstream/boot.html">Booting a Jetstream Computer Instance for your use!</a></li>
<li class="toctree-l1"><a class="reference internal" href="command-line.html">Intro to Shell Lesson</a></li>
<li class="toctree-l1"><a class="reference internal" href="command-line.html#the-shell">The Shell</a></li>
<li class="toctree-l1"><a class="reference internal" href="command-line.html#for-future-reference">For Future Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="command-line.html#finding-files">Finding files</a></li>
<li class="toctree-l1"><a class="reference internal" href="running-command-line-blast.html">Running command-line BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality.html">Short read quality and trimming</a></li>
<li class="toctree-l1"><a class="reference internal" href="assemble.html">Run the MEGAHIT assembler</a></li>
<li class="toctree-l1"><a class="reference internal" href="prokka_tutorial.html">Annotation with Prokka</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">A sourmash tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="#k-mers-k-mer-specificity-and-comparing-samples-with-k-mer-jaccard-distance">K-mers, k-mer specificity, and comparing samples with k-mer Jaccard distance.</a></li>
<li class="toctree-l1"><a class="reference internal" href="binning.html">Binning a Metagenomic Assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="salmon_tutorial.html">Gene Abundance Estimation with Salmon</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="slice.html">Slicing and dicing with k-mers</a></li>
<li class="toctree-l1"><a class="reference internal" href="anvio.html">Using Anvi&#8217;o to Knit Everything Together</a></li>
<li class="toctree-l1"><a class="reference internal" href="circos_tutorial.html">Using and Installing Circos</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Workflow and repeatability discussion</a></li>
<li class="toctree-l1"><a class="reference internal" href="DATA.html">Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="INSTALL.html">Building this site on your own</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html">2017 / April / Metagenomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="day2-install.html">Day 2 - installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="kmer_trimming.html">K-mer Spectral Error Trimming</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">A sourmash tutorial</a><ul>
<li><a class="reference internal" href="#objectives">Objectives</a></li>
<li><a class="reference internal" href="#at-the-beginning">At the beginning</a></li>
</ul>
</li>
<li><a class="reference internal" href="#k-mers-k-mer-specificity-and-comparing-samples-with-k-mer-jaccard-distance">K-mers, k-mer specificity, and comparing samples with k-mer Jaccard distance.</a><ul>
<li><a class="reference internal" href="#k-mers">K-mers!</a></li>
<li><a class="reference internal" href="#k-mers-and-assembly-graphs">K-mers and assembly graphs</a></li>
<li><a class="reference internal" href="#why-k-mers-though-why-not-just-work-with-the-full-read-sequences">Why k-mers, though? Why not just work with the full read sequences?</a></li>
<li><a class="reference internal" href="#long-k-mers-are-species-specific">Long k-mers are species specific</a></li>
<li><a class="reference internal" href="#using-k-mers-to-compare-samples-against-each-other">Using k-mers to compare samples against each other</a></li>
<li><a class="reference internal" href="#installing-sourmash">Installing sourmash</a></li>
<li><a class="reference internal" href="#generate-a-signature-for-illumina-reads">Generate a signature for Illumina reads</a></li>
<li><a class="reference internal" href="#compare-reads-to-assemblies">Compare reads to assemblies</a></li>
<li><a class="reference internal" href="#make-and-search-a-database-quickly">Make and search a database quickly.</a></li>
<li><a class="reference internal" href="#compare-many-signatures-and-build-a-tree">Compare many signatures and build a tree.</a></li>
<li><a class="reference internal" href="#what-s-in-my-metagenome">What&#8217;s in my metagenome?</a></li>
<li><a class="reference internal" href="#final-thoughts-on-sourmash">Final thoughts on sourmash</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="prokka_tutorial.html" title="Previous Chapter: Annotation with Prokka"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; Annotation wi...</span>
    </a>
  </li>
  <li>
    <a href="binning.html" title="Next Chapter: Binning a Metagenomic Assembly"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">Binning a Met... &raquo;</span>
    </a>
  </li>
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/sourmash.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">

    <div class="col-md-12 content">
      
  
  <div class="section" id="a-sourmash-tutorial">
<span id="a-sourmash-tutorial"></span><h1>A sourmash tutorial<a class="headerlink" href="#a-sourmash-tutorial" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="http://sourmash.readthedocs.io/en/latest/">sourmash</a> is our lab&#8217;s
implementation of an ultra-fast lightweight approach to
nucleotide-level search and comparison, called MinHash.</p>
<p>You can read some background about MinHash sketches in this paper:
<a class="reference external" href="http://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0997-x">Mash: fast genome and metagenome distance estimation using MinHash. Ondov BD, Treangen TJ, Melsted P, Mallonee AB, Bergman NH, Koren S, Phillippy AM. Genome Biol. 2016 Jun 20;17(1):132. doi: 10.1186/s13059-016-0997-x.</a></p>
<div class="section" id="objectives">
<span id="objectives"></span><h2>Objectives<a class="headerlink" href="#objectives" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Compare reads to assemblies</li>
<li>Create and search a custom database</li>
<li>Compare datasets and build a tree</li>
<li>Determine what&#8217;s in a metagenome (Taxonomic classification)</li>
</ul>
</div>
<div class="section" id="at-the-beginning">
<span id="at-the-beginning"></span><h2>At the beginning<a class="headerlink" href="#at-the-beginning" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="jetstream/boot.html">Create / log into</a> an m1.medium Jetstream instance,
and run these two commands:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>cd
curl -O https://s3-us-west-1.amazonaws.com/spacegraphcats.ucdavis.edu/microbe-genbank-sbt-k31-2017.05.09.tar.gz
tar xzf microbe-genbank-sbt-k31-2017.05.09.tar.gz
</pre></div>
</div>
<p>&#8211; they take a long time :).</p>
</div>
</div>
<div class="section" id="k-mers-k-mer-specificity-and-comparing-samples-with-k-mer-jaccard-distance">
<span id="k-mers-k-mer-specificity-and-comparing-samples-with-k-mer-jaccard-distance"></span><h1>K-mers, k-mer specificity, and comparing samples with k-mer Jaccard distance.<a class="headerlink" href="#k-mers-k-mer-specificity-and-comparing-samples-with-k-mer-jaccard-distance" title="Permalink to this headline">¶</a></h1>
<div class="section" id="k-mers">
<span id="k-mers"></span><h2>K-mers!<a class="headerlink" href="#k-mers" title="Permalink to this headline">¶</a></h2>
<p>K-mers are a fairly simple concept that turn out to be tremendously
powerful.</p>
<p>A &#8220;k-mer&#8221; is a word of DNA that is k long:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>ATTG - a 4-mer
ATGGAC - a 6-mer
</pre></div>
</div>
<p>Typically we extract k-mers from genomic assemblies or read data sets by
running a k-length window across all of the reads and sequences &#8211; e.g.
given a sequence of length 16, you could extract 11 k-mers of length six
from it like so:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">AGGATGAGACAGATAG</span>
</pre></div>
</div>
<p>becomes the following set of 6-mers:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>AGGATG
 GGATGA
  GATGAG
   ATGAGA
    TGAGAC
     GAGACA
      AGACAG
       GACAGA
        ACAGAT
         CAGATA
          AGATAG
</pre></div>
</div>
<p>k-mers are most useful when they&#8217;re <em>long</em>, because then they&#8217;re <em>specific</em>.
That is, if you have a 31-mer taken from a human genome, it&#8217;s pretty unlikely
that another genome has that exact 31-mer in it.  (You can calculate the
probability if you assume genomes are random: there are 4<sup>31</sup> possible
31-mers, and 4<sup>31</sup> = 4,611,686,018,427,387,904.  So, you know, a lot.)</p>
<p>The important concept here is that <strong>long k-mers are species specific</strong>.
We&#8217;ll go into a bit more detail later.</p>
</div>
<div class="section" id="k-mers-and-assembly-graphs">
<span id="k-mers-and-assembly-graphs"></span><h2>K-mers and assembly graphs<a class="headerlink" href="#k-mers-and-assembly-graphs" title="Permalink to this headline">¶</a></h2>
<p>We&#8217;ve already run into k-mers before, as it turns out - when we were
doing <a class="reference external" href="assemble.html">genome assembly</a>.  One of the three major
ways that genome assembly works is by taking reads, breaking them into
k-mers, and then &#8220;walking&#8221; from one k-mer to the next to bridge between
reads.  To see how this works, let&#8217;s take the 16-base sequence above,
and add another overlapping sequence:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>AGGATGAGACAGATAG
    TGAGACAGATAGGATTGC
</pre></div>
</div>
<p>One way to assemble these together is to break them down into k-mers &#8211;</p>
<p>becomes the following set of 6-mers:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>AGGATG
 GGATGA
  GATGAG
   ATGAGA
    TGAGAC
     GAGACA
      AGACAG
       GACAGA
        ACAGAT
         CAGATA
          AGATAG -&gt; off the end of the first sequence
           GATAGG &lt;- beginning of the second sequence
            ATAGGA
             TAGGAT
              AGGATT
               GGATTG
                GATTGC
</pre></div>
</div>
<p>and if you walk from one 6-mer to the next based on 5-mer overlap, you get
the assembled sequence:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">AGGATGAGACAGATAGGATTGC</span>
</pre></div>
</div>
<p>Graphs of many k-mers together are called De Bruijn graphs, and assemblers
like MEGAHIT and SOAPdenovo are De Bruijn graph assemblers - they use k-mers
underneath.</p>
</div>
<div class="section" id="why-k-mers-though-why-not-just-work-with-the-full-read-sequences">
<span id="why-k-mers-though-why-not-just-work-with-the-full-read-sequences"></span><h2>Why k-mers, though? Why not just work with the full read sequences?<a class="headerlink" href="#why-k-mers-though-why-not-just-work-with-the-full-read-sequences" title="Permalink to this headline">¶</a></h2>
<p>Computers <em>love</em> k-mers because there&#8217;s no ambiguity in matching them.
You either have an exact match, or you don&#8217;t.  And computers love that
sort of thing!</p>
<p>Basically, it&#8217;s really easy for a computer to tell if two reads share a
k-mer, and it&#8217;s pretty easy for a computer to store all the k-mers that
it sees in a pile of reads or in a genome.</p>
</div>
<div class="section" id="long-k-mers-are-species-specific">
<span id="long-k-mers-are-species-specific"></span><h2>Long k-mers are species specific<a class="headerlink" href="#long-k-mers-are-species-specific" title="Permalink to this headline">¶</a></h2>
<p>So, we&#8217;ve said long k-mers (say, k=31 or longer) are pretty species specific.
Is that really true?</p>
<p>Yes! Check out this figure from the <a class="reference external" href="http://msystems.asm.org/content/1/3/e00020-16">MetaPalette paper</a>:</p>
<p><img alt="x" src="_images/kmers-metapalette.png" /></p>
<p>here, the Koslicki and Falush show that k-mer similarity works to
group microbes by genus, at k=40. If you go longer (say k=50) then
you get only very little similarity between different species.</p>
</div>
<div class="section" id="using-k-mers-to-compare-samples-against-each-other">
<span id="using-k-mers-to-compare-samples-against-each-other"></span><h2>Using k-mers to compare samples against each other<a class="headerlink" href="#using-k-mers-to-compare-samples-against-each-other" title="Permalink to this headline">¶</a></h2>
<p>So, one thing you can do is use k-mers to compare genomes to genomes,
or read data sets to read data sets: data sets that have a lot of similarity
probably are similar or even the same genome.</p>
<p>One metric you can use for this comparisons is the Jaccard distance, which
is calculated by asking how many k-mers are <em>shared</em> between two samples
vs how many k-mers in total are in the combined samples.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>only k-mers in both samples
----------------------------
all k-mers in either or both samples
</pre></div>
</div>
<p>A Jaccard distance of 1 means the samples are identical; a Jaccard distance
of 0 means the samples are completely different.</p>
<p>This is a great measure and it can be used to search databases and
cluster unknown genomes and all sorts of other things!  The only real
problem with it is that there are a <em>lot</em> of k-mers in a genome &#8211;
a 5 Mbp genome (like <em>E. coli</em>) has 5 m k-mers!</p>
<p>About a year ago,
<a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0997-x">Ondov et al. (2016)</a>
showed that
<a class="reference external" href="https://en.wikipedia.org/wiki/MinHash">MinHash approaches</a> could be
used to estimate Jaccard distance using only a small fraction (1 in
10,000 or so) of all the k-mers.</p>
<p>The basic idea behind MinHash is that you pick a small subset of k-mers
to look at, and you use those as a proxy for <em>all</em> the k-mers.  The trick
is that you pick the k-mers randomly but consistently: so if a chosen
k-mer is present in two data sets of interest, it will be picked in both.
This is done using a clever trick that we can try to explain to you in
class - but either way, trust us, it works!</p>
<p>We have implemented a MinHash approach in our
<a class="reference external" href="https://github.com/dib-lab/sourmash/">sourmash software</a>, which can
do some nice things with samples.  We&#8217;ll show you some of these things
next!</p>
</div>
<div class="section" id="installing-sourmash">
<span id="installing-sourmash"></span><h2>Installing sourmash<a class="headerlink" href="#installing-sourmash" title="Permalink to this headline">¶</a></h2>
<p>To install sourmash, run:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sudo apt-get -y update &amp;&amp; \
sudo apt-get install -y python3.5-dev python3.5-venv make \
    libc6-dev g++ zlib1g-dev
</pre></div>
</div>
<p>this installs Python 3.5.</p>
<p>Now, create a local software install and populate it with Jupyter and
other dependencies:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>
python3.5 -m venv ~/py3
. ~/py3/bin/activate
pip install -U pip
pip install -U Cython
pip install -U jupyter jupyter_client ipython pandas matplotlib scipy scikit-learn khmer

pip install -U https://github.com/dib-lab/sourmash/archive/master.zip
</pre></div>
</div>
</div>
<div class="section" id="generate-a-signature-for-illumina-reads">
<span id="generate-a-signature-for-illumina-reads"></span><h2>Generate a signature for Illumina reads<a class="headerlink" href="#generate-a-signature-for-illumina-reads" title="Permalink to this headline">¶</a></h2>
<p>Download some reads and a reference genome:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>mkdir ~/data
cd ~/data
wget http://public.ged.msu.edu.s3.amazonaws.com/ecoli_ref-5m-trim.pe.fq.gz
wget https://s3.amazonaws.com/public.ged.msu.edu/ecoliMG1655.fa.gz
</pre></div>
</div>
<p>Compute a scaled MinHash signature from our reads:</p>
<p><img alt="" src="_images/Sourmash_flow_diagrams_QC.png" />
<img alt="" src="_images/Sourmash_flow_diagrams_compute.png" /></p>
<div class="highlight-python"><div class="highlight"><pre><span></span>mkdir ~/sourmash
cd ~/sourmash

sourmash compute --scaled 10000 ~/data/ecoli_ref*pe*.fq.gz -o ecoli-reads.sig -k 31
</pre></div>
</div>
</div>
<div class="section" id="compare-reads-to-assemblies">
<span id="compare-reads-to-assemblies"></span><h2>Compare reads to assemblies<a class="headerlink" href="#compare-reads-to-assemblies" title="Permalink to this headline">¶</a></h2>
<p>Use case: how much of the read content is contained in the reference genome?</p>
<p>Build a signature for an <em>E. coli</em> genome:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sourmash compute --scaled 10000 -k 31 ~/data/ecoliMG1655.fa.gz -o ecoli-genome.sig
</pre></div>
</div>
<p>and now evaluate <em>containment</em>, that is, what fraction of the read content is
contained in the genome:</p>
<p><img alt="" src="_images/Sourmash_flow_diagrams_search.png" /></p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sourmash search -k 31 ecoli-reads.sig ecoli-genome.sig --containment
</pre></div>
</div>
<p>and you should see:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span># running sourmash subcommand: search
loaded query: /home/ubuntu/data/ecoli_ref-5m... (k=31, DNA)
loaded 1 signatures from ecoli-genome.sig
1 matches:
similarity   match
----------   -----
 46.6%       /home/ubuntu/data/ecoliMG1655.fa.gz
</pre></div>
</div>
<p>Why are only 50% or so of our k-mers from the reads in the genome!?
Any ideas?</p>
<p>Try the reverse - why is it bigger?</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sourmash search -k 31 ecoli-genome.sig ecoli-reads.sig --containment
</pre></div>
</div>
<p>(...but 99% of our k-mers from the genome are in the reads!?)</p>
<p>This is basically because of sequencing error! Illumina data contains
a lot of errors, and the assembler doesn&#8217;t include them in the assembly!</p>
</div>
<div class="section" id="make-and-search-a-database-quickly">
<span id="make-and-search-a-database-quickly"></span><h2>Make and search a database quickly.<a class="headerlink" href="#make-and-search-a-database-quickly" title="Permalink to this headline">¶</a></h2>
<p>Suppose that we have a collection of signatures (made with <code class="docutils literal"><span class="pre">sourmash</span> <span class="pre">compute</span></code> as above) and we want to search it with our newly assembled
genome (or the reads, even!). How would we do that?</p>
<p>Let&#8217;s grab a sample collection of 50 E. coli genomes and unpack it &#8211;</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>mkdir ecoli_many_sigs
cd ecoli_many_sigs

curl -O -L https://github.com/dib-lab/sourmash/raw/master/data/eschericia-sigs.tar.gz

tar xzf eschericia-sigs.tar.gz
rm eschericia-sigs.tar.gz

cd ../
</pre></div>
</div>
<p>This will produce 50 files named <code class="docutils literal"><span class="pre">ecoli-N.sig</span></code> in the <code class="docutils literal"><span class="pre">ecoli_many_sigs</span></code> &#8211;</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>ls ecoli_many_sigs
</pre></div>
</div>
<p>Let&#8217;s turn this into an easily-searchable database with <code class="docutils literal"><span class="pre">sourmash</span> <span class="pre">index</span></code> &#8211;</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sourmash index -k 31 ecolidb ecoli_many_sigs/*.sig
</pre></div>
</div>
<p>What does the database look like and how does the search work?</p>
<p><img alt="" src="_images/SBT.png" /></p>
<p>sourmash search (and gather) compare the query and the database (sequence bloom tree)
and report matches based containment.</p>
<p>One point to make with this is that the search can quickly narrow down
which signatures match your query, without losing any matches.  It&#8217;s a
clever example of how computer scientists can actually make life
better :).</p>
<hr class="docutils" />
<p>And now we can search!</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sourmash search ecoli-genome.sig ecolidb.sbt.json -n 20
</pre></div>
</div>
<p>You should see output like this:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span># running sourmash subcommand: search
select query k=31 automatically.
loaded query: /home/tx160085/data/ecoliMG165... (k=31, DNA)
loaded SBT ecolidb.sbt.json
Searching SBT ecolidb.sbt.json
49 matches; showing first 20:
similarity   match
----------   -----
 75.4%       NZ_JMGW01000001.1 Escherichia coli 1-176-05_S4_C2 e117605...
 72.2%       NZ_GG774190.1 Escherichia coli MS 196-1 Scfld2538, whole ...
 71.4%       NZ_JMGU01000001.1 Escherichia coli 2-011-08_S3_C2 e201108...
 70.1%       NZ_JHRU01000001.1 Escherichia coli strain 100854 100854_1...
 69.0%       NZ_JH659569.1 Escherichia coli M919 supercont2.1, whole g...
 64.9%       NZ_JNLZ01000001.1 Escherichia coli 3-105-05_S1_C1 e310505...
 63.0%       NZ_MOJK01000001.1 Escherichia coli strain 469 Cleandata-B...
 62.9%       NZ_MOGK01000001.1 Escherichia coli strain 676 BN4_676_1_(...
 62.0%       NZ_JHDG01000001.1 Escherichia coli 1-176-05_S3_C1 e117605...
 59.9%       NZ_MIWF01000001.1 Escherichia coli strain AF7759-1 contig...
 52.7%       NZ_KE700241.1 Escherichia coli HVH 147 (4-5893887) acYxy-...
 51.7%       NZ_APWY01000001.1 Escherichia coli 178200 gec178200.conti...
 49.3%       NZ_LVOV01000001.1 Escherichia coli strain swine72 swine72...
 49.3%       NZ_MIWP01000001.1 Escherichia coli strain K6412 contig_00...
 49.0%       NZ_LQWB01000001.1 Escherichia coli strain GN03624 GCID_EC...
 48.9%       NZ_JHGJ01000001.1 Escherichia coli O45:H2 str. 2009C-4780...
 48.1%       NZ_CP011331.1 Escherichia coli O104:H4 str. C227-11, comp...
 47.7%       NZ_JHNB01000001.1 Escherichia coli O103:H25 str. 2010C-45...
 47.7%       NZ_JHRE01000001.1 Escherichia coli strain 302014 302014_1...
 47.6%       NZ_JHHE01000001.1 Escherichia coli O103:H2 str. 2009C-327...
</pre></div>
</div>
<p>identifying what genome is in the signature.</p>
</div>
<div class="section" id="compare-many-signatures-and-build-a-tree">
<span id="compare-many-signatures-and-build-a-tree"></span><h2>Compare many signatures and build a tree.<a class="headerlink" href="#compare-many-signatures-and-build-a-tree" title="Permalink to this headline">¶</a></h2>
<p>Adjust plotting (this is a bug in sourmash :) &#8211;</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>echo &#39;backend : Agg&#39; &gt; matplotlibrc
</pre></div>
</div>
<p>Compare all the signatures:</p>
<p><img alt="" src="_images/Sourmash_flow_diagrams_compare.png" /></p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sourmash compare ecoli_many_sigs/* -o ecoli_cmp
</pre></div>
</div>
<p>and then plot:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sourmash plot --labels ecoli_cmp
</pre></div>
</div>
<p>which will produce a file <code class="docutils literal"><span class="pre">ecoli_cmp.matrix.png</span></code> and <code class="docutils literal"><span class="pre">ecoli_cmp.dendro.png</span></code>
which you can then download via FileZilla and view on your local computer.</p>
<p>Now open jupyter notebook and visualize:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">Image</span>
<span class="n">Image</span><span class="p">(</span><span class="s2">&quot;ecoli_cmp.matrix.png&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Here&#8217;s a PNG version:</p>
<p><img alt="E. coli comparison plot" src="_images/ecoli_cmp.matrix.png" /></p>
</div>
<div class="section" id="what-s-in-my-metagenome">
<span id="what-s-in-my-metagenome"></span><h2>What&#8217;s in my metagenome?<a class="headerlink" href="#what-s-in-my-metagenome" title="Permalink to this headline">¶</a></h2>
<p>At the beginning, we downloaded and unpacked a GenBank index of all
the microbial genomes &#8211; you can see a basic description here,
<a class="reference external" href="http://ivory.idyll.org/blog/2016-sourmash-sbt-more.html">CTB&#8217;s blog post</a>
&#8211; this one contains sketches of all 100k Genbank microbes. (See
<a class="reference external" href="http://sourmash.rtfd.io/en/latest/databases.html">available sourmash databases</a>
for more information.)</p>
<p>After this database is unpacked, it produces a file
<code class="docutils literal"><span class="pre">genbank-k31.sbt.json</span></code> and a whole bunch of hidden files in the
directory <code class="docutils literal"><span class="pre">.sbt.genbank-k31</span></code>.</p>
<p>Next, run the &#8216;gather&#8217; command to see what&#8217;s in your ecoli genome &#8211;</p>
<p><img alt="" src="_images/Sourmash_flow_diagrams_gather.png" /></p>
<p>at the command line run:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sourmash gather -k 31 ecoli-reads.sig ../genbank-k31.sbt.json -o ecoli-reads-gather-reults.csv
</pre></div>
</div>
<p>and you should get:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span># running sourmash subcommand: sbt_gather
loaded query: /home/tx160085/data/ecoli_ref-... (k=31, DNA)
loaded SBT ../genbank-k31.sbt.json

overlap     p_query p_match
---------   ------- --------
4.9 Mbp      46.6%  100.0%      APIN01000001.1 Escherichia coli str. ...
3.4 Mbp      32.4%    1.5%      AQAX01000001.1 Escherichia coli P0304...
2.5 Mbp      23.6%    0.7%      JHDK01000001.1 Escherichia coli 1-110...
160.0 kbp     1.5%    0.6%      AMPE01000001.1 Citrobacter sp. L17 co...
0.5 Mbp       4.8%    0.4%      BBVS01000001.1 Escherichia albertii D...
3.3 Mbp      31.2%    0.4%      APYP01000001.1 Escherichia coli P0299...
3.6 Mbp      34.5%    0.4%      AQCO01000001.1 Escherichia coli MP021...
4.3 Mbp      40.9%    0.4%      JONB01000001.1 Escherichia coli 3-020...
2.8 Mbp      26.0%    0.5%      AZPL01000001.1 Shigella flexneri 2001...
2.8 Mbp      26.0%    0.2%      MIIY01000001.1 Shigella sp. FC1737 NO...
2.2 Mbp      20.7%    0.2%      AFET01000004.1 Escherichia coli AA86 ...
found less than 10.0 kbp in common. =&gt; exiting

found 11 matches total;
the recovered matches hit 49.2% of the query
</pre></div>
</div>
<p>You can use this on metagenomes (assembled and
unassembled) as well; you&#8217;ve just got to make the signature files.</p>
<p>To see this in action, here is gather running on a signature generated
from some sequences that assemble (but don&#8217;t align to known genomes)
from the
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/23387867">Shakya et al. 2013 mock metagenome paper</a>.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>wget https://github.com/dib-lab/sourmash/raw/master/doc/_static/shakya-unaligned-contigs.sig
sourmash gather -k 31 shakya-unaligned-contigs.sig ../genbank-k31.sbt.json
</pre></div>
</div>
<p>This should yield:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span># running sourmash subcommand: gather
loaded query: mqc500.QC.AMBIGUOUS.99.unalign... (k=31, DNA)
loaded SBT genbank-k31.sbt.json

overlap     p_query p_match
---------   ------- --------
1.4 Mbp      11.0%   58.0%      JANA01000001.1 Fusobacterium sp. OBRC1 c
1.0 Mbp       7.7%   25.9%      CP001957.1 Haloferax volcanii DS2 plasmi
0.9 Mbp       7.5%   11.8%      BA000019.2 Nostoc sp. PCC 7120 DNA, comp
0.7 Mbp       5.9%   23.0%      FOVK01000036.1 Proteiniclasticum ruminis
0.7 Mbp       5.3%   17.6%      AE017285.1 Desulfovibrio vulgaris subsp.
0.6 Mbp       4.9%   11.1%      CP001252.1 Shewanella baltica OS223, com
0.6 Mbp       4.8%   27.3%      AP008226.1 Thermus thermophilus HB8 geno
0.6 Mbp       4.4%   11.2%      CP000031.2 Ruegeria pomeroyi DSS-3, comp
480.0 kbp     3.8%    7.6%      CP000875.1 Herpetosiphon aurantiacus DSM
410.0 kbp     3.3%   10.5%      CH959317.1 Sulfitobacter sp. NAS-14.1 sc
1.4 Mbp      10.9%   11.8%      LN831027.1 Fusobacterium nucleatum subsp
0.5 Mbp       4.1%    5.3%      CP000753.1 Shewanella baltica OS185, com
420.0 kbp     3.3%    7.7%      FNDZ01000023.1 Proteiniclasticum ruminis
150.0 kbp     1.2%    4.5%      CP015081.1 Deinococcus radiodurans R1 ch
150.0 kbp     1.2%    8.2%      CP000969.1 Thermotoga sp. RQ2, complete
290.0 kbp     2.3%    4.1%      CH959311.1 Sulfitobacter sp. EE-36 scf_1
1.2 Mbp       9.4%    5.0%      CP013328.1 Fusobacterium nucleatum subsp
110.0 kbp     0.9%    3.5%      FREL01000833.1 Enterococcus faecalis iso
0.6 Mbp       5.0%    2.8%      CP000527.1 Desulfovibrio vulgaris DP4, c
340.0 kbp     2.7%    3.3%      KQ235732.1 Fusobacterium nucleatum subsp
70.0 kbp      0.6%    1.2%      CP000850.1 Salinispora arenicola CNS-205
60.0 kbp      0.5%    0.7%      CP000270.1 Burkholderia xenovorans LB400
50.0 kbp      0.4%    2.6%      CP001080.1 Sulfurihydrogenibium sp. YO3A
50.0 kbp      0.4%    3.2%      L77117.1 Methanocaldococcus jannaschii D
found less than 40.0 kbp in common. =&gt; exiting

found 24 matches total;
the recovered matches hit 73.4% of the query
</pre></div>
</div>
<p>In our
<a class="reference external" href="biorxiv.org/content/biorxiv/early/2017/06/25/155358.full.pdf">recent preprint</a>
using this, we showed that</p>
<p>It is straightforward to build your own databases for use with
<code class="docutils literal"><span class="pre">search</span></code> and <code class="docutils literal"><span class="pre">gather</span></code>; this is of interest if you have dozens or
hundreds of sequencing data sets in your group. Ping us if you want us
to write that up.</p>
</div>
<div class="section" id="final-thoughts-on-sourmash">
<span id="final-thoughts-on-sourmash"></span><h2>Final thoughts on sourmash<a class="headerlink" href="#final-thoughts-on-sourmash" title="Permalink to this headline">¶</a></h2>
<p>There are many tools like Kraken and Kaiju that can do taxonomic
classification of individual reads from metagenomes; these seem to
perform well (albeit with high false positive rates) in situations
where you don&#8217;t necessarily have the genome sequences that are in the
metagenome.  Sourmash, by contrast, can estimate which known genomes are
actually present, so that you can extract them and map/align to them.
It seems to have a very low false positive rate and is quite sensitive
to strains.</p>
<p>Above, we&#8217;ve shown you a few things that you can use sourmash for.  Here
is a (non-exclusive) list of other uses that we&#8217;ve been thinking about &#8211;</p>
<ul class="simple">
<li>detect contamination in sequencing data;</li>
<li>index and search private sequencing collections;</li>
<li>search all of SRA for overlaps in metagenomes;</li>
</ul>
<p>Chat with Luiz, Phil, or Titus if you are interested in these use cases!</p>
</div>
</div>



<hr>

<font size="-1"><b>LICENSE:</b>
This documentation and all textual/graphic site content is released
under <a href='http://creativecommons.org/publicdomain/zero/1.0/'>
Creative Commons - 0
(CC0)</a> -- <a href='https://github.com/ngs-docs/2017-ucsc-metagenomics'>fork @
github</a>. </font>

<hr>


<!-- disqus commenting disabled; set disqus_shortname -->

    </div>
    

  </div>
</div>

<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2017, Lab for Data Intensive Biology.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.<br/>
    </p>
  </div>
</footer>




<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', '', 'auto');
  ga('send', 'pageview');

</script>


  </body>
</html>